[CDAP-13380] TMSLogAppender and MessagingLogsProcessorService

anwar6953 opened the pull request:
https://issues.cask.co/browse/CDAP-13380
Implemented:
TMSLogAppender, which publishes log messages to TMS.
MessagingLogsProcessorService, which subscribes from TMS and appends the logs to a LocalLogAppender.

Commit:
Improve erroneous test case.

Commit:
Publish logs to TMS on standalone.

rohitsinha54 commented:
It will be good to ask for review with at least a single/small unit test which shows the core piece of change works. If there is no unit test then how do we know if the code actually work.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,267 @@
+      this.logPartitionType =
+              LogPartitionType.valueOf(cConf.get(Constants.Logging.LOG_PUBLISH_PARTITION_KEY).toUpperCase());
+      this.messagingContext = new MultiThreadMessagingContext(messagingService);
+//      this.stringPartitioner = new StringPartitioner(cConf);
```
why not remove this?

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,267 @@
+
+    private void serializeMessageToBuffer(LogMessage logMessage, Map<Integer, List<byte[]>> buffer) {
+      String partitionKey = logPartitionType.getPartitionKey(logMessage.getLoggingContext());
+//      int partition = stringPartitioner.partition(partitionKey, numPartitions);
```
why not remove it?

Commit:
Append RuntimeMonitor's monitored log topics to a configured LogAppender.
Added test case for TMSLogAppender.
Miscellaneous other cleanup, such as removal of unused code paths.

CuriousVini commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -1345,7 +1345,7 @@
       instances, each CDAP instance needs to distinguish its Kafka topics
       from those of other CDAP instances with unique values for
-      ${log.kafka.topic} and ${metrics.topic.prefix}.
+      ${log.kafka.topic}, log.tms.topic, and ${metrics.topic.prefix}.
```
Is this change right? should it be ${log.tms.topic.prefix}

anwar6953 replied:
yep; modified. thanks.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/MessagingLogsProcessorService.java:
```diff
@@ -0,0 +1,205 @@
+
+  @Override
+  protected void run() {
+    LOG.info("Start running MessagingLogsProcessorService");
```
should we use `Starting {} service, service.getName()`?

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/MessagingLogsProcessorService.java:
```diff
@@ -0,0 +1,205 @@
+
+  @Override
+  protected void triggerShutdown() {
+    LOG.info("Stopping Logs Processing Service.");
```
same here: should we use `Stopping {} service, service.getName()`?

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/MessagingLogsProcessorService.java:
```diff
@@ -0,0 +1,205 @@
+    Futures.getUnchecked(Futures.allAsList(futures));
+
+    localLogAppender.stop();
+    LOG.info("Logs Processing Service stopped.");
```
same here. It will be good to use same name.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/MessagingLogsProcessorService.java:
```diff
@@ -0,0 +1,205 @@
+
+    List<ListenableFuture<State>> futures =
+            processLogsServices.stream().map(AbstractScheduledService::start).collect(Collectors.toList());
+    Futures.getUnchecked(Futures.allAsList(futures));
```
why not log here when service has already started?

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java:
```diff
@@ -104,6 +101,8 @@ private KafkaLogPublisher(CConfiguration cConf, BlockingQueue<LogMessage> messag
       this.loggingEventSerializer = new LoggingEventSerializer();
       this.contextAware = contextAware;
       this.retryStrategy = RetryStrategies.fromConfiguration(cConf, "system.log.process.");
+      this.logPartitionType =
```
do we still need this class if we are using TMS for publishing messages?

anwar6953 replied:
Yes, we still want to partition across the TMS topics for parallelism.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,264 @@
+    protected void triggerShutdown() {
+      // Interrupt the run thread first
+      // If the run loop is sleeping / blocking, it will wake and break the loop
+      Thread runThread = this.blockingThread;
```
why do we need another variable `runThread`?

anwar6953 replied:
I think its because otherwise, after the null check, another thread may change the value to null.
I modeled this after https://github.com/caskdata/cdap/blob/develop/cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,264 @@
+ */
+public final class TMSLogAppender extends LogAppender {
+
+  private static final int QUEUE_SIZE = 512;
```
this should be configurable

anwar6953 replied:
done.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,264 @@
+  @Override
+  public void start() {
+    tmsLogPublisher.startAndWait();
+    addInfo("Successfully initialized TMSLogAppender.");
```
did you mean successfully started?

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,264 @@
+
+  @Override
+  public void stop() {
+    tmsLogPublisher.stopAndWait();
```
Also log here

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,264 @@
+      // Publish all remaining messages.
+      while (!messageQueue.isEmpty() || !buffer.isEmpty()) {
+        try {
+          publishMessages(buffer, false);
```
why does this not require transactional publishing?

anwar6953 replied:
I think that for logs, we don't need it. I believe logs are best-effort.
There is a chance that some logs will be sent duplicated, if publishing to TMS fails and is retried, but that is the tradeoff for avoiding transactional overhead for logs.

Maybe @chtyim or @anew can give their input.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,264 @@
+        bufferSize += partitionBuffer.getValue().size();
+      }
+
+      if (blockForMessage) {
```
what is the purpose of this?

anwar6953 replied:
Most of the time, there is no point in going past this because there are no messages to publish yet.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,264 @@
+            throws IOException, InterruptedException, TopicNotFoundException {
+      int bufferSize = 0;
+      for (Map.Entry<Integer, List<byte[]>> partitionBuffer : buffer.entrySet()) {
+        bufferSize += partitionBuffer.getValue().size();
```
why dont we check the limit on buffrsize here?

anwar6953 replied:
This is just computing the partition buffer size, to do checks against it later.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,264 @@
+     * @param buffer a buffer for storing {@code byte[]} for publishing to TMS
+     * @throws InterruptedException if the thread is interrupted
+     */
+    private void publishMessages(Map<Integer, List<byte[]>> buffer, boolean blockForMessage)
```
can you add implementation approach in this method in detail in the javadoc?

anwar6953 replied:
improved the javadoc.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/MessagingLogsProcessorService.java:
```diff
@@ -0,0 +1,205 @@
+/**
+ * Processes log events stored in TMS.
+ */
+public class MessagingLogsProcessorService extends AbstractExecutionThreadService {
```
The name should be LocalMessagingLogsProcessorService since it will only be used for standalone. Also add that to the Javadoc

CuriousVini reviewed (changes requested):
1.) Please add unit tests for both `TMSLogAppender` and `MessagingLogsProcessorService`
2.) Please update license year wherever applicable

anwar6953 commented:
Removed MessagingLogsProcessorService.java in favor of a RemoteExecutionLogProcessor.java.

anwar6953 commented:
Added unit tests and addressed comments.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,267 @@
+
+      int failures = 0;
+      long failureStartTime = System.currentTimeMillis();
+      while (isRunning()) {
```
Extends from `AbstractRetryableScheduledService` instead to simplify the looping and retry logic.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java:
```diff
@@ -0,0 +1,267 @@
+/**
+ * Log appender that publishes log messages to TMS.
+ */
+public final class TMSLogAppender extends LogAppender {
```
This class has certain amount of duplicated code as the `KafkaLogAppender`. Please refactor it to have a common abstract class and abstract out the actual publish logic. Also, please use `AbstractRetryableScheduledService` instead of `AbstractExecutionThreadService` to simplify the logic.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java:
```diff
@@ -276,9 +286,22 @@ private long processResponse(Map<String, Deque<MonitorMessage>> monitorResponses
-    messagePublisher.publish(NamespaceId.SYSTEM.getNamespace(), topic,
-                             messages.stream().map(MonitorMessage::getMessage).iterator());
+    if (topic.startsWith(cConf.get(Constants.Logging.TMS_TOPIC_PREFIX))) {
+      messages.forEach(monitorMessage -> {
```
It's better to move this logic into the `RemoteExecutionLogProcessor` implementation, especially for the Kafka case, we should't need to deserialize and reserialize the object.

Commit:
Address comments:
Refactor TMSLogAppender and KafkaLogAppender to reuse code.
For retryies, use AbstractRetryableScheduledService.
Move log message deserialization logic into RemoteExecutionLogProcessor.

anwar6953 commented:
Build is green: https://builds.cask.co/browse/CDAP-DUT6503

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/remote/RemoteExecutionLogProcessor.java:
```diff
@@ -0,0 +1,30 @@
+   *
+   * @param loggingEventBytes the event coming from the remote execution
+   */
+  void process(byte[] loggingEventBytes);
```
Better have it takes `Iterator<byte[]>` to allow batching in Kafka case. In the caller, you can do:

`logProcessor.process(monitorMessages.stream().map(MonitorMessage::getMessage).iterator())`

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/app/runtime/monitor/RuntimeMonitorTest.java:
```diff
@@ -91,6 +92,7 @@
   private DatasetService datasetService;
   private DatasetFramework datasetFramework;
   private Transactional transactional;
+  private RemoteExecutionLogProcessor noOpLogProcessor;
```
instead of having a field, it's actually easier to read to do it in-place.

`new RuntimeMonitor(...., m -> { });`

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/service/AbstractRetryableScheduledService.java:
```diff
@@ -133,6 +133,16 @@ protected final void shutDown() throws Exception {
+   * Called when the RetryStrategy used by this class returns a negative number, indicating to abort the operation.
+   *
+   * @param e the exception which exhausted retries
+   * @return the number of milliseconds to delay until the next call to this method
```
This is not accurate. It is the number of ms to delay until the next iterate gets triggered again.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/service/AbstractRetryableScheduledService.java:
```diff
@@ -133,6 +133,16 @@ protected final void shutDown() throws Exception {
   }
 
+  /**
+   * Called when the RetryStrategy used by this class returns a negative number, indicating to abort the operation.
```
This is not clear what's the purpose of this method is. This method is give the chance for sub-class to reset the retry count or to terminate the service (by throwing exception).

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/service/AbstractRetryableScheduledService.java:
```diff
@@ -153,7 +163,9 @@ protected final void runOneIteration() throws Exception {
                                                                     failureCount,
                                                                     System.currentTimeMillis() - nonFailureStartTime)));
-        throw e;
+        delayMillis = handleRetriesExhausted(e);
```
Should take a `Maths.max(0L, handleRetriesExhausted(e))` to guard against -ve number.

chtyim commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -1345,7 +1345,7 @@
       instances, each CDAP instance needs to distinguish its Kafka topics
       from those of other CDAP instances with unique values for
-      ${log.kafka.topic} and ${metrics.topic.prefix}.
+      ${log.kafka.topic}, ${log.tms.topic}, and ${metrics.topic.prefix}.
```
We don't have this for kafka right??

chtyim commented on cdap-master/src/main/java/co/cask/cdap/data/tools/HBaseQueueDebugger.java:
```diff
@@ -499,6 +500,7 @@ private static Injector createInjector(boolean disableAuthorization) throws Exce
       new ProgramRunnerRuntimeModule().getDistributedModules(),
       new SystemDatasetRuntimeModule().getDistributedModules(),
       new NotificationServiceRuntimeModule().getDistributedModules(),
+      new LoggingModules().getDistributedModules(),
```
How does it work before?

anwar6953 replied:
The bindings for the RemoteExecutionTwillRunnerService class did not previously have any dependencies on any logging bindings, but now it does (due to RuntimeMonitor depending on logging modules' dependencies).

See the failing test cases here: https://builds.cask.co/browse/CDAP-DUT6503-4/test

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+ *
+ * @param <BUFFER_MESSAGE> the type of message used in the in-memory buffer, before publishing
+ */
+public abstract class LogPublisher<BUFFER_MESSAGE> extends AbstractRetryableScheduledService {
```
Call it `AbstractLogPublisher`.

chtyim replied:
Also the param should just be `MESSAGE`.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+    super(retryStrategy);
+    this.queueSize = queueSize;
+    this.messageQueue = messageQueue;
+    this.contextAware = contextAware;
```
Instead of taking a `ContextAware` class, better just leave the `handleRetriesExhausted` for the children classes to implement. Otherwise you'll be forcing sub-class to provide an implementation of `ContextAware`. Alternatively, have an method `logError` for the children to implement / override.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+    return 0;
+  }
+
+    @Override
```
Misalign

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+  private final BlockingQueue<LogMessage> messageQueue;
+  private final ContextAware contextAware;
+  private final List<BUFFER_MESSAGE> buffer;
+  private volatile boolean hadFailure;
```
just call it `failed`.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+    return super.shouldRetry(ex);
+  }
+
+    @Override
```
misalign.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+  private final ContextAware contextAware;
+  private final List<BUFFER_MESSAGE> buffer;
+  private volatile boolean hadFailure;
+  private volatile Thread blockingThread;
```
Call this `publishThread`.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+      runThread.interrupt();
+    }
+
+    // Publish all remaining messages.
```
You should clear the interrupt of the current thread by calling `Thread.interrupted()`

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+
+  protected abstract BUFFER_MESSAGE createMessage(LogMessage logMessage);
+
+  protected abstract void publish(List<BUFFER_MESSAGE> logMessages) throws Exception;
```
Missing javadoc.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+    // Log using the status manager
+//  contextAware.addError("Failed to publish log message to Kafka on topic " + topic, e);
+
+  protected abstract BUFFER_MESSAGE createMessage(LogMessage logMessage);
```
Missing javadoc.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+  }
+
+    // Log using the status manager
+//  contextAware.addError("Failed to publish log message to Kafka on topic " + topic, e);
```
Remove.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+      buffer.clear();
+    }
+
+    super.doShutdown();
```
no need.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java:
```diff
@@ -0,0 +1,141 @@
+                      ContextAware contextAware) {
+    super(retryStrategy);
+    this.queueSize = queueSize;
+    this.messageQueue = messageQueue;
```
Why the message queue is coming from child class? Generally not a good idea to share collection. I think this class should own the queue and expose a `protected final` method for sub-class to call to add `LogMessage`.

chtyim commented on cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/tms/TestTMSLogging.java:
```diff
@@ -0,0 +1,180 @@
+
+    Logger logger = LoggerFactory.getLogger("TestTMSLogging");
+    LoggingTester loggingTester = new LoggingTester();
+    FlowletLoggingContext flowletLoggingContext = new FlowletLoggingContext("TKL_NS_1", "APP_1", "FLOW_1", "FLOWLET_1",
```
Don't use Flowlet for new code.

anwar6953 commented:
@chtyim Addressed comments; please take another look.
Thanks.

chtyim commented on cdap-master/src/main/java/co/cask/cdap/data/tools/HBaseQueueDebugger.java:
```diff
@@ -499,6 +500,7 @@ private static Injector createInjector(boolean disableAuthorization) throws Exce
       new ProgramRunnerRuntimeModule().getDistributedModules(),
       new SystemDatasetRuntimeModule().getDistributedModules(),
       new NotificationServiceRuntimeModule().getDistributedModules(),
+//      new LoggingModules().getDistributedModules(),
```
Why commented?

anwar6953 replied:
oops - I commented it to make the test fail, to understand why it was needed. Forgot to uncomment it.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/AbstractLogPublisher.java:
```diff
@@ -0,0 +1,169 @@
+   * @param logMessage the log message to enqueue for publishing
+   * @throws InterruptedException if interrupted while waiting to enqueue (waiting is done in case the queue is full)
+   */
+  public final void enqueueForPublishing(LogMessage logMessage) throws InterruptedException {
```
Just all it `addMessage`. Calling it enqueue kind of exposing the implementation.

chtyim reviewed (commented):
Just one comment about the naming.

anwar6953 commented:
Build is green: https://builds.cask.co/browse/CDAP-DUT6503-7

Squashed the 'addressing comments' commits, and merging.
